grpc/http/forwarders.go

```diff
@@ -133,6 +143,75 @@ type StreamForwarderFunc func(
 	opts ...func(context.Context, http.ResponseWriter, proto.Message) error,
 )
 
+func writeKeepalive(w http.ResponseWriter, mut *sync.Mutex) {
+	mut.Lock()
+	defer mut.Unlock()
+
+	// Per https://html.spec.whatwg.org/multipage/server-sent-events.html#event-stream-interpretation,
+	// lines that start with a `:` must be ignored by the client.
+	_, err := w.Write([]byte(":\n"))
+
+	if err != nil {
+		log.Warnf("failed to write http keepalive response: %v", err)
+	} else if f, ok := w.(http.Flusher); ok {
+		f.Flush()
+	}
+}
+
+func keepalive(ctx context.Context, w http.ResponseWriter, mut *sync.Mutex, wg *sync.WaitGroup, t ticker.Ticker) {
+	defer t.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-t.C():
+			writeKeepalive(w, mut)
+
+			// The waitgroup is purely intended for unit tests and is always nil in production use cases
+			if wg != nil {
+				wg.Done()
+			}
+		}
+	}
+}
+
+func withKeepalive(ctx context.Context, w http.ResponseWriter, recv recvFn) (http.ResponseWriter, recvFn) {
+	return withKeepaliveAux(ctx, w, recv, nil, ticker.NewTicker)
+}
+
+func withKeepaliveAux(ctx context.Context, w http.ResponseWriter, recv recvFn, wg *sync.WaitGroup, tf tickerFac) (http.ResponseWriter, recvFn) {
+	keepaliveInterval := time.Duration(time.Second * 15)
+	keepaliveTicker := tf(keepaliveInterval)
+	oldRecv := recv
+	mut := sync.Mutex{}
+	oldw := w
+
+	w = httpsnoop.Wrap(oldw, httpsnoop.Hooks{
```

Comment:
I ended up still needing `httpsnoop` so that I could implement the lock in the right spot

Reply:
questions - can we simplify this by getting rid of the tests? I know that sounds mad, but it is hard to tests concurrent code, and I think it is usually better not to have unit tests and try and avoid bugs by making the code as simple as possible. 

AFAIK, you need to start a goroutine with the response and sent a keep-alive every 30s. I would not worry about resetting it either.

Reply:
If I remove the unit tests, I can definitely simplify this a bit since I would not need to be concerned about mocking out the timer, and I would not need to pass in a waitgroup.  I'm happy to remove the code to reset the timer as well if you feel that's simpler - that would eliminate the extra wrapper around the `recv` fn.  `httpsnoop` is still necessary unfortunately in order to take a lock when writing to the responsewriter since we could have multiple goroutines attempting to do so.

Reply:
Actually, we can even keep the `reset` since I can move it to the wrapped `Write` fn if we want.  Since the keepalive function receives the original, unwrapped ResponseWriter, we can easily tell which writes should reset the timer and which ones shouldn't.

Reply:
I went ahead and made the changes discussed, removing the tests and simplifying the code a bit. I was able to maintain the resetting the ticker like I mentioned without complicating the code at all since we already need to wrap the ResponseWriter for the mutex.  Let me know what you think